### PR TITLE
SSO implementation

### PR DESCRIPTION
**Pull request by yashkrishan**
# SSO Authentication with Account Linking

## Overview

This PR implements a comprehensive SSO (Single Sign-On) authentication system that enables users to sign in with work email credentials (Google Workspace, Microsoft Azure AD) while maintaining full backward compatibility with existing GitHub OAuth. The system prevents account duplication through email-based identity management and provides a seamless account linking experience.

### 1. Multi-Provider Authentication
- **Google Workspace SSO** - Sign in with Google work accounts
- **Microsoft Azure AD SSO** - Sign in with Microsoft work accounts
- **GitHub OAuth** - Existing functionality preserved (migrated to new system)
- **Extensible Architecture** - Base provider class for adding new SSO providers

### 2. Smart Account Linking
- **Email-Based Identity** - Email serves as universal identifier across all providers
- **Zero Duplication** - Database constraints prevent duplicate accounts
- **User Confirmation Flow** - Secure linking process when connecting multiple providers
- **Provider Management** - Users can link/unlink providers and set primary provider

### 3. Security & Compliance
- **Audit Logging** - Complete authentication event tracking
- **CSRF Protection** - State parameters and secure tokens in OAuth flows
- **Organization Support** - Domain-based organization extraction and SSO configuration
- **15-Minute Expiring Tokens** - Secure account linking confirmation

### 4. Backward Compatibility
- All existing users automatically migrated
- Existing `/api/v1/login` and `/api/v1/signup` endpoints unchanged
- Firebase authentication continues to work
- Zero breaking changes

#### Database Schema
- **4 New Tables:**
  - `user_auth_providers` - Multiple authentication providers per user
  - `pending_provider_links` - Temporary storage for account linking confirmation
  - `organization_sso_config` - Organization-level SSO settings
  - `auth_audit_log` - Security audit trail for all authentication events
- **12 Indexes** for query optimization
- **Migration Script** - Automatically migrates existing users to new schema

#### New Services
- **UnifiedAuthService** (`unified_auth_service.py`) - 631 lines
  - Multi-provider authentication logic
  - Account linking with user confirmation
  - Provider management (add, remove, set primary)
  - Organization domain detection
  - Audit logging integration

#### SSO Providers
- **GoogleProvider** (`google_provider.py`) - Google Workspace SSO
- **AzureProvider** (`azure_provider.py`) - Microsoft Azure AD SSO
- **BaseSSOProvider** (`base_provider.py`) - Extensible base class

#### API Endpoints
- `POST /api/v1/sso/login` - SSO authentication
- `POST /api/v1/providers/confirm-linking` - Confirm account linking
- `DELETE /api/v1/providers/cancel-linking/{token}` - Cancel linking request
- `GET /api/v1/providers/me` - List user's authentication providers
- `POST /api/v1/providers/set-primary` - Set primary provider
- `DELETE /api/v1/providers/unlink` - Unlink a provider
- `GET /api/v1/account/me` - Get complete account information

### Frontend Changes (potpie-ui)

- Updated auth pages (`sign-in`, `sign-up`, `onboarding`)
- New SSO authentication components
- Account linking dialog implementation
- Removed Firebase SSO setup documentation (migrated to new system)

### Testing

- **60+ Unit Tests** covering:
  - UnifiedAuthService (40+ tests)
  - Database models (20+ tests)
  - Auth router endpoints
- **Integration Test Placeholders** for end-to-end flows
- **Manual Testing Guide** included

### Documentation

- **API Documentation** (`API_DOCUMENTATION.md`) - Complete endpoint reference
- **Backend Implementation Guide** - Service layer documentation
- **Frontend Implementation Guide** - UI integration guide
- **Implementation Summary** - Feature overview
- **Code Examples** - Usage examples for all scenarios
- **Manual Testing Guide** - Step-by-step testing procedures

### For Existing Users
- All existing users are automatically migrated on first database migration
- GitHub OAuth preserved as `firebase_github` provider
- Set as primary provider automatically
- **No action required from users**

### For Developers
- Existing authentication endpoints remain unchanged
- New SSO endpoints run in parallel
- Can be enabled gradually via feature flags
- Database migration must be run: `alembic upgrade head`

### Unit Tests
```bash
pytest app/modules/auth/tests/
```

### Test Coverage
-  All authentication scenarios (new user, existing user, account linking)
- Provider management operations
- Error handling and edge cases
- Database constraints and validations

##  Statistics

- **39 files changed**
- **11,203 insertions**, 92 deletions
- **4 new database tables**
- **7 new API endpoints**
- **60+ unit tests**
- **6 comprehensive documentation files**

##  Checklist

- [x] Database migration tested and verified
- [x] All unit tests passing
- [x] Backward compatibility maintained
- [x] Documentation complete
- [x] Code reviewed
- [x] Security audit logging implemented
- [x] Error handling comprehensive
- [x] Manual testing completed

---

**Branch:** `feat/sso-authentication-with-account-linking`  
**Base:** `main`  
**Status:** Ready for Review

## Summary by CodeRabbit

* **New Features**
  * Google SSO and multi‑provider auth: SSO login, link/confirm/cancel provider linking, set primary, unlink, and full account view.
  * Organization detection from email domains and per‑organization SSO configuration.
  * SSO provider registry and standardized SSO provider interface.

* **Security**
  * Encryption of stored provider tokens and audit logging for auth events.

* **Tests**
  * New unit/integration tests and pytest configuration for auth/SSO flows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: app/modules/auth/auth_schema.py [hunk 1]
```diff
@@ -1,6 +1,185 @@
...
+
+
+# ===== SSO Login Flow =====
+
+
+class SSOLoginRequest(BaseModel):
+    """Request to login via SSO provider"""
+
+    email: str = Field(..., description="User's work email")
+    sso_provider: str = Field(..., description="'google', 'azure', 'okta', 'saml'")
```
Comment (Dsantra92) [inline app/modules/auth/auth_schema.py:70]: Use enum maybe?

File: app/modules/auth/tests/test_auth_router.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+
+class TestSSOLoginEndpoint:
+    """Test POST /api/v1/sso/login"""
+
+    @pytest.mark.skip("Requires full FastAPI app setup")
+    def test_sso_login_new_user(self, client, db_session):
+        """Test SSO login for new user creates account"""
+        response = client.post(
+            "/api/v1/sso/login",
```
Comment (Dsantra92) [inline app/modules/auth/tests/test_auth_router.py:31]: We should ideally be using `/api/v2`.

File: app/modules/code_provider/github/github_service.py [hunk 3]
```diff
@@ -181,9 +181,128 @@ def _detect_encoding(content_bytes: bytes) -> str:
...
+                ):
+                    other_access_token = other_user.provider_info.get("access_token")
+                    if other_access_token:
+                        logger.info(
+                            f"Found GitHub token in provider_info for user "
+                            f"{other_user.uid} (same email as {uid}). Using it."
+                        )
+                        return other_access_token
+        except Exception as e:
+            logger.debug(f"Error checking other users with same email: {str(e)}")
```
Comment (Dsantra92) [inline app/modules/code_provider/github/github_service.py]: Don't use f strings in logging.

File: app/modules/users/user_service.py [hunk 4]
```diff
@@ -135,7 +135,7 @@ def get_user_id_by_email(self, email: str) -> str:
             logger.error(f"DEBUG: Error fetching user ID by email {email}: {e}")
             return None
 
-    async def get_user_by_email(self, email: str) -> User:
+    def get_user_by_email(self, email: str) -> Optional[User]:
```
Comment (Dsantra92) [inline app/modules/users/user_service.py]: Have we handled all of the places where it expects async?

Comment (yashkrishan) [reply app/modules/users/user_service.py]: yes

File: app/modules/code_provider/github/github_service.py (hunk 3 above)
Comment (yashkrishan) [reply app/modules/code_provider/github/github_service.py]: ok

File: app/modules/auth/auth_provider_model.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    provider_uid = Column(
+        String(255), nullable=False
+    )  # Provider's unique ID for this user
+    provider_data = Column(JSONB)  # Full provider response/metadata
+
+    # OAuth tokens (for API access, e.g., GitHub API)
+    access_token = Column(
+        Text
+    )  # OAuth access token (should be encrypted in production)
+    refresh_token = Column(Text)  # OAuth refresh token (should be encrypted)
```
[the remark this answers is not in the document]
Comment (dhirenmathur) [reply app/modules/auth/auth_provider_model.py]: We should do this wherever tokens are being used and stored. This is required for SOC 2 as well

File: app/modules/auth/sso_providers/azure_provider.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
+"""
```
Comment (dhirenmathur) [inline app/modules/auth/sso_providers/azure_provider.py]: Can we get rid of all Azure-related files since we're not using it? We can always get it back from Git history and I don't want to commit dead code into the repository

File: app/modules/auth/unified_auth_service.py [hunk 7]
```diff
@@ -0,0 +1,644 @@
...
+
+    def __init__(self, db: Session):
+        self.db = db
+        self.user_service = UserService(db)
+
+        # Initialize SSO providers
+        self.sso_providers: Dict[str, BaseSSOProvider] = {
+            "google": GoogleSSOProvider(),
+            "azure": AzureSSOProvider(),
+        }
```
Comment (dhirenmathur) [inline app/modules/auth/unified_auth_service.py]: Can we do a singleton here instead of initializing it on every class init

**Review by dhirenmathur — CHANGES_REQUESTED** [review]

File: app/modules/auth/sso_providers/azure_provider.py (hunk 6 above)
Comment (yashkrishan) [reply app/modules/auth/sso_providers/azure_provider.py]: done

File: app/modules/auth/unified_auth_service.py (hunk 7 above)
Comment (yashkrishan) [reply app/modules/auth/unified_auth_service.py]: ok

**Review by dhirenmathur — APPROVED** [review]
